File: skill-lib/msdmd/parsers/universal.ts
// 110:15
/**
 * Universal msdmd parser — pure Node stdlib (fs, path).
 *
 * TypeScript counterpart to parsers/universal.py. Implements the
 * parser contract from msdmd/SKILL.md: extracts every
 * `// === <BLOCK_NAME> ===` … `// === END <BLOCK_NAME> ===` block
 * from a source file and returns its entries as flat objects.
 *
 * Comment marker auto-detected by file extension. The block syntax
 * itself is identical across languages; only the per-line marker
 * changes.
 *
 * Zero non-stdlib dependencies. Safe to copy verbatim into any
 * Node/Deno/Bun project that wants msdmd support.
 */
import { readFileSync, statSync, readdirSync } from "node:fs";
import { join, extname } from "node:path";

export type Entry = Record<string, string>;

const MARKERS: Record<string, string> = {
  ".py": "#", ".rb": "#", ".ex": "#", ".exs": "#", ".sh": "#",
  ".ts": "//", ".tsx": "//", ".js": "//", ".jsx": "//", ".mjs": "//",
  ".rs": "//", ".go": "//", ".java": "//", ".c": "//", ".cpp": "//",
  ".cc": "//", ".h": "//", ".hpp": "//", ".swift": "//", ".kt": "//",
  ".sql": "--", ".lua": "--", ".hs": "--",
};

const DEFAULT_SKIP = new Set([
  "__pycache__", "node_modules", ".git", ".venv", "venv",
  "dist", "build", ".next", ".nuxt", "target", ".pytest_cache",
  ".mypy_cache", ".tox",
]);

function escapeRegex(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

export function markerFor(path: string): string | null {
  return MARKERS[extname(path).toLowerCase()] ?? null;
}

export function parseText(
  text: string,
  blockName: string,
  marker: string = "#",
): Entry[] {
  const m = escapeRegex(marker);
  const name = escapeRegex(blockName);
  const blockRe = new RegExp(
    `^${m} === ${name} ===\\s*$([\\s\\S]*?)^${m} === END ${name} ===\\s*$`,
    "gm",
  );
  const idRe = new RegExp(`^\\s*${m}\\s*id:\\s*(\\S+)\\s*$`);
  const fieldRe = new RegExp(`^\\s*${m}\\s+([a-z_]+):\\s*(.+?)\\s*$`);

  const entries: Entry[] = [];
  let match: RegExpExecArray | null;
  while ((match = blockRe.exec(text)) !== null) {
    const body = match[1];
    let current: Entry | null = null;
    for (const rawLine of body.split("\n")) {
      const line = rawLine.replace(/\s+$/, "");
      const mid = idRe.exec(line);
      if (mid) {
        if (current !== null) entries.push(current);
        current = { id: mid[1] };
        continue;
      }
      if (current === null) continue;
      const mf = fieldRe.exec(line);
      if (mf) current[mf[1]] = mf[2];
    }
    if (current !== null) entries.push(current);
  }
  return entries;
}

export function parseFile(path: string, blockName: string): Entry[] {
  const marker = markerFor(path);
  if (marker === null) return [];
  try {
    return parseText(readFileSync(path, "utf-8"), blockName, marker);
  } catch {
    return [];
  }
}

export interface WalkOptions {
  skip?: Set<string>;
  extensions?: Set<string>;
}

export function walkTree(
  root: string,
  blockName: string,
  opts: WalkOptions = {},
): { annotated: Array<[string, Entry[]]>; untested: string[] } {
  const skip = opts.skip ?? DEFAULT_SKIP;
  const extensions =
    opts.extensions ?? new Set(Object.keys(MARKERS));

  const annotated: Array<[string, Entry[]]> = [];
  const untested: string[] = [];

  function visit(dir: string): void {
    let names: string[];
    try {
      names = readdirSync(dir).sort();
    } catch {
      return;
    }
    for (const name of names) {
      if (skip.has(name)) continue;
      const full = join(dir, name);
      let st;
      try {
        st = statSync(full);
      } catch {
        continue;
      }
      if (st.isDirectory()) {
        visit(full);
      } else if (st.isFile()) {
        if (!extensions.has(extname(full).toLowerCase())) continue;
        const entries = parseFile(full, blockName);
        if (entries.length > 0) {
          annotated.push([full, entries]);
        } else {
          untested.push(full);
        }
      }
    }
  }

  visit(root);
  return { annotated, untested };
}
// 110:15

File: skill-lib/msdmd/parsers/universal.py
# 85:29
"""Universal msdmd parser — pure stdlib.

Implements the parser contract from ``msdmd/SKILL.md``: extracts every
``# === <BLOCK_NAME> ===`` … ``# === END <BLOCK_NAME> ===`` block from
a source file and returns its entries as flat dicts.

Comment marker is auto-detected by file extension. The block syntax
itself is identical across languages; only the per-line marker changes.

Public API:

    parse_text(text, block_name, marker="#") -> list[dict]
    parse_file(path, block_name) -> list[dict]
    walk_tree(root, block_name, *, skip=None) -> tuple[annotated, untested]

This module has zero non-stdlib dependencies and is safe to copy
verbatim into any project that wants msdmd support.
"""
from __future__ import annotations
import re
from pathlib import Path
from typing import Iterable

# extension → comment marker
_MARKERS: dict[str, str] = {
    ".py": "#", ".rb": "#", ".ex": "#", ".exs": "#", ".sh": "#",
    ".ts": "//", ".tsx": "//", ".js": "//", ".jsx": "//", ".mjs": "//",
    ".rs": "//", ".go": "//", ".java": "//", ".c": "//", ".cpp": "//",
    ".cc": "//", ".h": "//", ".hpp": "//", ".swift": "//", ".kt": "//",
    ".sql": "--", ".lua": "--", ".hs": "--",
}

_DEFAULT_SKIP = (
    "__pycache__", "node_modules", ".git", ".venv", "venv",
    "dist", "build", ".next", ".nuxt", "target", ".pytest_cache",
    ".mypy_cache", ".tox",
)


def marker_for(path: Path) -> str | None:
    """Return the comment marker for a file path, or None if unsupported."""
    return _MARKERS.get(path.suffix.lower())


def _block_regex(block_name: str, marker: str) -> re.Pattern[str]:
    m = re.escape(marker)
    name = re.escape(block_name)
    return re.compile(
        rf"^{m} === {name} ===\s*$(?P<body>.*?)^{m} === END {name} ===\s*$",
        re.MULTILINE | re.DOTALL,
    )


def parse_text(text: str, block_name: str, marker: str = "#") -> list[dict]:
    """Extract every entry from every matching block in ``text``.

    Entries are flat ``dict[str, str]`` keyed by field name. The first
    line of an entry must be ``id: <value>``; subsequent lines until
    the next ``id:`` (or block end) carry indented ``<key>: <value>``
    pairs.
    """
    block_re = _block_regex(block_name, marker)
    m = re.escape(marker)
    id_re = re.compile(rf"^\s*{m}\s*id:\s*(?P<id>\S+)\s*$")
    field_re = re.compile(rf"^\s*{m}\s+(?P<key>[a-z_]+):\s*(?P<val>.+?)\s*$")

    entries: list[dict] = []
    for block in block_re.finditer(text):
        current: dict[str, str] | None = None
        for line in block.group("body").splitlines():
            line = line.rstrip()
            mid = id_re.match(line)
            if mid:
                if current is not None:
                    entries.append(current)
                current = {"id": mid.group("id")}
                continue
            if current is None:
                continue
            mf = field_re.match(line)
            if mf:
                current[mf.group("key")] = mf.group("val")
        if current is not None:
            entries.append(current)
    return entries


def parse_file(path: Path, block_name: str) -> list[dict]:
    """Parse a single file. Returns [] if the file's extension has no
    known comment marker or if the file can't be read."""
    marker = marker_for(path)
    if marker is None:
        return []
    try:
        return parse_text(path.read_text(encoding="utf-8"), block_name, marker)
    except (OSError, UnicodeDecodeError):
        return []


def walk_tree(
    root: Path,
    block_name: str,
    *,
    skip: Iterable[str] | None = None,
    extensions: Iterable[str] | None = None,
) -> tuple[list[tuple[Path, list[dict]]], list[Path]]:
    """Walk ``root`` and partition source files into (annotated, untested).

    ``annotated`` is a list of ``(path, entries)`` for every file that
    contains at least one entry of ``block_name``. ``untested`` is every
    other source file (still filtered by extension and skip-dirs) so
    coverage gaps remain observable.
    """
    skip_set = set(skip) if skip is not None else set(_DEFAULT_SKIP)
    ext_set = (
        set(e.lower() if e.startswith(".") else "." + e.lower() for e in extensions)
        if extensions is not None
        else set(_MARKERS.keys())
    )
    annotated: list[tuple[Path, list[dict]]] = []
    untested: list[Path] = []
    for path in sorted(root.rglob("*")):
        if not path.is_file():
            continue
        if any(part in skip_set for part in path.parts):
            continue
        if path.suffix.lower() not in ext_set:
            continue
        entries = parse_file(path, block_name)
        if entries:
            annotated.append((path, entries))
        else:
            untested.append(path)
    return annotated, untested
# 85:29
